From d923ea19992a0b4450e648f38eda1190bab0eda6 Mon Sep 17 00:00:00 2001
From: Lina Iyer <lina.iyer@linaro.org>
Date: Mon, 9 Nov 2015 12:47:32 -0700
Subject: [PATCH 18/25] drivers: cpu-pd: Add next_wakeup to device's timing
 data

Allow devices that know when their next wakeup event is, to record save
it as part of timing data. A genpd governor may use this data to
determine if suspending the domain is going to affect the QoS of its
devices.

Add API to read the next event for the CPU. The value can be evaluated
at any time to determine the remaining sleep time of the CPU. Record the
next wake up of each CPU when the CPU suspends.

Cc: Thomas Gleixner <tglx@linutronix.de>
Signed-off-by: Lina Iyer <lina.iyer@linaro.org>
---
 drivers/base/power/cpu-pd.c | 14 ++++++++++++++
 include/linux/pm_domain.h   |  2 ++
 include/linux/tick.h        | 10 ++++++++++
 kernel/time/tick-sched.c    |  8 ++++++++
 4 files changed, 34 insertions(+)

diff --git a/drivers/base/power/cpu-pd.c b/drivers/base/power/cpu-pd.c
index d19c5af..8ec2bf3 100644
--- a/drivers/base/power/cpu-pd.c
+++ b/drivers/base/power/cpu-pd.c
@@ -20,6 +20,7 @@
 #include <linux/rculist.h>
 #include <linux/rcupdate.h>
 #include <linux/slab.h>
+#include <linux/tick.h>
 
 #define CPU_PD_NAME_MAX 36
 
@@ -30,6 +31,13 @@ void cpu_pm_runtime_suspend(void)
 {
 	struct device *cpu_dev = get_cpu_device(smp_processor_id());
 
+	if (cpu_dev->pm_domain) {
+		struct generic_pm_domain_data *gpd;
+
+		gpd = dev_gpd_data(cpu_dev);
+		gpd->td.next_wakeup = tick_nohz_get_next_wakeup();
+	}
+
 	RCU_NONIDLE(__pm_runtime_suspend(cpu_dev, 0));
 }
 EXPORT_SYMBOL(cpu_pm_runtime_suspend);
@@ -43,6 +51,12 @@ void cpu_pm_runtime_resume(void)
 
 	RCU_NONIDLE(__pm_runtime_resume(cpu_dev, 0));
 
+	if (cpu_dev->pm_domain) {
+		struct generic_pm_domain_data *gpd;
+
+		gpd = dev_gpd_data(cpu_dev);
+		gpd->td.next_wakeup.tv64 = 0;
+	}
 }
 EXPORT_SYMBOL(cpu_pm_runtime_resume);
 
diff --git a/include/linux/pm_domain.h b/include/linux/pm_domain.h
index aa91659..237da5d 100644
--- a/include/linux/pm_domain.h
+++ b/include/linux/pm_domain.h
@@ -16,6 +16,7 @@
 #include <linux/of.h>
 #include <linux/notifier.h>
 #include <linux/spinlock.h>
+#include <linux/ktime.h>
 
 /* Defines used for the flags field in the struct generic_pm_domain */
 #define GENPD_FLAG_PM_CLK	(1U << 0) /* PM domain uses PM clk */
@@ -104,6 +105,7 @@ struct gpd_timing_data {
 	s64 effective_constraint_ns;
 	bool constraint_changed;
 	bool cached_stop_ok;
+	ktime_t next_wakeup;
 };
 
 struct pm_domain_data {
diff --git a/include/linux/tick.h b/include/linux/tick.h
index e312219..6db3e52 100644
--- a/include/linux/tick.h
+++ b/include/linux/tick.h
@@ -103,6 +103,7 @@ extern void tick_nohz_idle_enter(void);
 extern void tick_nohz_idle_exit(void);
 extern void tick_nohz_irq_exit(void);
 extern ktime_t tick_nohz_get_sleep_length(void);
+extern ktime_t tick_nohz_get_next_wakeup(void);
 extern u64 get_cpu_idle_time_us(int cpu, u64 *last_update_time);
 extern u64 get_cpu_iowait_time_us(int cpu, u64 *last_update_time);
 #else /* !CONFIG_NO_HZ_COMMON */
@@ -116,6 +117,15 @@ static inline ktime_t tick_nohz_get_sleep_length(void)
 
 	return len;
 }
+
+static inline ktime_t tick_nohz_get_next_wakeup(void)
+{
+	ktime_t len = { .tv64 = NSEC_PER_SEC/HZ };
+
+	/* Next wake up is the tick period, assume it starts now */
+	return ktime_add(len, ktime_get());
+}
+
 static inline u64 get_cpu_idle_time_us(int cpu, u64 *unused) { return -1; }
 static inline u64 get_cpu_iowait_time_us(int cpu, u64 *unused) { return -1; }
 #endif /* !CONFIG_NO_HZ_COMMON */
diff --git a/kernel/time/tick-sched.c b/kernel/time/tick-sched.c
index 11ce599..40f77af 100644
--- a/kernel/time/tick-sched.c
+++ b/kernel/time/tick-sched.c
@@ -870,6 +870,14 @@ ktime_t tick_nohz_get_sleep_length(void)
 	return ts->sleep_length;
 }
 
+ktime_t tick_nohz_get_next_wakeup(void)
+{
+	struct clock_event_device *dev =
+			__this_cpu_read(tick_cpu_device.evtdev);
+
+	return dev->next_event;
+}
+
 static void tick_nohz_account_idle_ticks(struct tick_sched *ts)
 {
 #ifndef CONFIG_VIRT_CPU_ACCOUNTING_NATIVE
-- 
1.9.1

